cluster-autoscaler/core/scale_up.go

```diff
@@ -35,9 +35,129 @@ import (
 	"k8s.io/autoscaler/cluster-autoscaler/utils/nodegroupset"
 	schedulercache "k8s.io/kubernetes/pkg/scheduler/cache"
 
+	"fmt"
 	"github.com/golang/glog"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"math"
 )
 
+type scaleUpResourcesLimits map[string]int64
+type scaleUpResourcesDelta map[string]int64
+
+// used as a value in scaleUpResourcesLimits if actual limit could not be obtained due to errors talking to cloud provider
+const scaleUpLimitUnknown = math.MaxInt64
+
+func computeScaleUpResourcesLeftLimits(nodeGroups []cloudprovider.NodeGroup, nodeInfos map[string]*schedulercache.NodeInfo, resourceLimiter *cloudprovider.ResourceLimiter) scaleUpResourcesLimits {
+	totalCores, totalMem, errCoresMem := calculateScaleUpCoresMemoryTotal(nodeGroups, nodeInfos)
+
+	resultScaleUpLimits := make(scaleUpResourcesLimits)
+	for _, resource := range resourceLimiter.GetResources() {
+		max := resourceLimiter.GetMax(resource)
+
+		// we put only actual limits into final map. No entry means no limit.
+		if max > 0 {
+			switch {
+			case resource == cloudprovider.ResourceNameCores:
+				if errCoresMem != nil {
+					resultScaleUpLimits[resource] = scaleUpLimitUnknown
+				} else {
+					resultScaleUpLimits[resource] = computeBelowMax(totalCores, max)
+				}
+
+			case resource == cloudprovider.ResourceNameMemory:
+				if errCoresMem != nil {
+					resultScaleUpLimits[resource] = scaleUpLimitUnknown
+				} else {
+					resultScaleUpLimits[resource] = computeBelowMax(totalMem, max)
+				}
+
+			default:
+				glog.Errorf("Scale up limits defined for unsupported resource '%s'", resource)
+			}
+		}
+	}
+	return resultScaleUpLimits
+}
+
+func calculateScaleUpCoresMemoryTotal(nodeGroups []cloudprovider.NodeGroup, nodeInfos map[string]*schedulercache.NodeInfo) (int64, int64, error) {
+	var coresTotal int64
+	var memoryTotal int64
+	failed := false
+
+	for _, nodeGroup := range nodeGroups {
+		currentSize, err := nodeGroup.TargetSize()
+		if err != nil {
+			glog.Errorf("Failed to get node group size of %v: %v", nodeGroup.Id(), err)
+			failed = true
+			continue
+		}
+		nodeInfo, found := nodeInfos[nodeGroup.Id()]
+		if !found {
+			glog.Errorf("No node info for: %s", nodeGroup.Id())
+			failed = true
+			continue
+		}
+		if currentSize > 0 {
+			nodeCPU, nodeMemory := getNodeInfoCoresAndMemory(nodeInfo)
+			coresTotal = coresTotal + int64(currentSize)*nodeCPU
+			memoryTotal = memoryTotal + int64(currentSize)*nodeMemory
+		}
+	}
+
+	if failed {
```

Comment:
Is there any reason for setting `failed` to true and returning error here as opposed to immediately returning error?

Reply:
More logging I guess.